client/service/src/builder.rs

```diff
@@ -740,6 +741,46 @@ ServiceBuilder<
 		+ TransactionPoolMaintainer<Block=TBl, Hash = <TBl as BlockT>::Hash>,
 	TRpc: sc_rpc::RpcExtension<sc_rpc::Metadata> + Clone,
 {
+
+	/// Set an ExecutionExtensionsMaker
+	pub fn with_execution_extensions_maker(self, execution_extensions_maker: Box<dyn ExtensionsMaker>) -> Result<Self, Error> {
```

Comment:
This is an ugly API.

A different way to provide this would be to go through `service.client` e.g.:

```rust
		builder.client().execution_extensions().set_extensions_maker(Box::new(()));
```
I am open to discussing that as the "official" way to do this in 2.0 for now and maybe mark `builder.client()` as unstable. Which would make clear that if we changed how this works internally, this might break your code!

Reply:
I'd prefer going through `client` to be honest, you can do that after the service is actually built, no? Like this is something you can change at any point if you have a reference to client anyway, so why try to restrict it to service builder anyway?

Reply:
I don't understand why this is an ugly api? This just follows the builder pattern? Or do I miss something?

Reply:
> I'd prefer going through client to be honest, you can do that after the service is actually built, no? Like this is something you can change at any point if you have a reference to client anyway, so why try to restrict it to service builder anyway?

Being able to do this after the service is built seems like a huge hack to me that we should explicitly forbid for the sake of it.

Reply:
Agree with @tomaka here. The _only_ reason the set_extensions_maker is wrapped in a `RwLock` is because the order in which the setup happens (first create the client, later call build – not the other way round) prevents us from doing it in a clean way–same for  the transaction-pool extensions. Changing _either_ after construction / during the run (which this makes possible) makes no sense and will have unexepected and unwanted side-effects. Just think of having different Externalities or Transaction Pools between two Exection-Callls ... really not expected behavior.

It's an API that I dearly would  like to see removed, thus I'd prefer to not leak it too wide or recommend its usage. Hence the wrapper we have here...

@bkchr it's ugly because internally and externally it _doesn't_ match the pattern around: it isn't accepting a `FnOnce` (nor `Fn`) like the others do; it isn't executed immediately and the result stored and it isn't even kept on the builder but directly set on client...

Reply:
> It's an API that I dearly would like to see removed, thus I'd prefer to not leak it too wide or recommend its usage. Hence the wrapper we have here...

What about adding a comment about that in the code?
From my experience, if we don't explicitly mention to not do something, sooner or later someone will do a small PR against that design without understanding the implications.

Reply:
@tomaka I created a ticket and added a FIXME in the code.